relay-config/src/config.rs

```diff
@@ -3,47 +3,63 @@ use std::fmt;
 use std::fs;
 use std::io;
 use std::io::Write;
-use std::net::{IpAddr, SocketAddr, ToSocketAddrs};
+use std::net::{AddrParseError, IpAddr, SocketAddr, ToSocketAddrs};
 use std::path::{Path, PathBuf};
 use std::time::Duration;
 
 use failure::{Backtrace, Context, Fail};
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
 
-use relay_auth::{generate_key_pair, generate_relay_id, PublicKey, RelayId, SecretKey};
-use relay_common::Dsn;
+use relay_auth::{
+    generate_key_pair, generate_relay_id, KeyParseError, PublicKey, RelayId, SecretKey,
+};
+use relay_common::{Dsn, Uuid};
 use relay_redis::RedisConfig;
 
 use crate::types::ByteSize;
 use crate::upstream::UpstreamDescriptor;
+use crate::UpstreamParseError;
 
 macro_rules! ctry {
     ($expr:expr, $kind:expr, $path:expr) => {
         match $expr {
             Ok(val) => val,
-            Err(err) => return Err(ConfigError::new($path, err.context($kind))),
+            Err(err) => return Err(ConfigError::new_file_error($path, err.context($kind))),
         }
     };
 }
 
+/// Defines the source of a config error
+#[derive(Debug)]
+enum ErrorSource {
```

Comment:
We already have a `ConfigErrorKind` with values like CouldNotOpenFile, BadYaml and InvalidOverride ... for all except InvalidOverride the source is a configuration file, for InvalidOverride the source is an Environment Variable or a CLI parameter. The Error source is defined to keep details specific to the source of error (i.e. file or env var).

Reply:
Gotcha. `source` may be a bit misleading because of [`Error::source`](https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.source). Let's leave it for now, though 👍